### PR TITLE
Provider.request with injected providers

### PR DESCRIPTION
**Pull request by Velenir**
Now that **web3** supports `request` method, this PR

- Adds **LogMiddleware** to debug `JsonRpcRequest|Response`, on by default in develop, disabled with `window.logMiddleware(false)`
- Enhances **composedProvider**, to handle`request` method if available on injected provider

Need to make it work with **WalletConnect** next

Please test with **Metamaks** (worked for me great) and dapp browsers

Part of #1199

### CODE REVIEW

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr1232--dexreact.review.gnosisdev.com

File: src/api/wallet/composeProvider.ts [hunk 1]
```diff
@@ -5,12 +5,44 @@ import RpcEngine, {
...
 import { web3 } from 'api'
+import { createLoggerMiddleware } from './loggerMiddleware'
+
+type RpcCallBack<T extends unknown> = (error: JsonRpcError<T>, res: JsonRpcResponse<T>) => void
+
+function providerFromEngine<T extends Provider>(engine: JsonRpcEngine): T {
+  const sendAsync = engine.handle.bind(engine)
+
+  const send = <T extends unknown>(req: JsonRpcRequest<T>, callback: RpcCallBack<T>): void => {
+    if (!callback) throw new Error('Web3 Provider - must provider callback to "send" method')
```
Comment (alfetopito) [inline src/api/wallet/composeProvider.ts:21]:
```suggestion
    if (!callback) throw new Error('Web3 Provider - must provide callback to "send" method')
```

**Comment by Velenir** [comment]
Will switch logging off by default before merging

**Comment by Velenir** [comment]
If it's slow with logging, can you try `window.logMiddleware(false)`

File: src/api/wallet/composeProvider.ts [hunk 2]
```diff
@@ -5,12 +5,44 @@ import RpcEngine, {
...
+  const sendAsync = engine.handle.bind(engine)
+
+  const send = <T extends unknown>(req: JsonRpcRequest<T>, callback: RpcCallBack<T>): void => {
+    if (!callback) throw new Error('Web3 Provider - must provide callback to "send" method')
+    engine.handle(req, callback)
+  }
+  const request = <T extends unknown>(req: JsonRpcRequest<T>): Promise<JsonRpcResponse<T>['result']> => {
+    return new Promise((resolve, reject) => {
+      engine.handle(req, (error: JsonRpcError<T>, res: JsonRpcResponse<T>) => {
+        // console.log('CPROV::handled error:', error, 'res:', res)
```
Comment (W3stside) [inline src/api/wallet/composeProvider.ts:27]: dead code

**Comment by anxolin** [comment]
open

**Comment by anxolin** [comment]
Superseded by https://github.com/gnosis/gp-ui/pull/8

**Pull request closed without merging**